File: src/main/java/net/sf/rails/game/state/HashSetState.java
package net.sf.rails.game.state;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * State class that wraps a HashSet
 */
public final class HashSetState<E> extends SetState<E> {

    private final HashSet<E> set;

    private HashSetState(Item parent, String id, Collection<E> collection) {
        super(parent, id);
        if (collection == null) {
            set = new HashSet<E>();
        } else {
            set = new HashSet<E>(collection);
        }
    }

    /**
     * @return empty HashSetState
     */
    public static <E> HashSetState<E> create(Item parent, String id){
        return new HashSetState<E>(parent, id, null);
    }
    
    /**
     * @return prefilled HashSetState
     */
    public static <E> HashSetState<E> create(Item parent, String id, Collection<E> collection){
        return new HashSetState<E>(parent, id, collection);
    }
    
    @Override
    protected Set<E> getSet() {
        return set;
    }
}

File: dotnet/RailsLib.Net/Net/Game/State/HashSetState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameLib.Net.Game.State
{
    public class HashSetState<T> : SetState<T>
    {
        private readonly HashSet<T> set;

        private HashSetState(IItem parent, string id, IEnumerable<T> collection) : base(parent, id)
        {
            if (collection == null)
            {
                set = new HashSet<T>();
            }
            else
            {
                set = new HashSet<T>(collection);
            }
        }

        /**
         * @return empty HashSetState
         */
        public static HashSetState<T> Create(IItem parent, string id)
        {
            return new HashSetState<T>(parent, id, null);
        }

        /**
         * @return prefilled HashSetState
         */
        public static HashSetState<T> Create(IItem parent, string id, IEnumerable<T> collection)
        {
            return new HashSetState<T>(parent, id, collection);
        }

        override public ISet<T> GetSet()
        {
            return set;
        }
    }
}
